### PR TITLE
Build arguments

### PR DESCRIPTION
**Pull request by stevenwilliamson**
- Adds a --noclean option that prevents the auto clean up of tempdirs
  used during the build. This means if a gem won't build you can inspect
  the log files.
- adds a --build-config option that allows you to specify config to be
  used when building the gem. For example to set where to look for
  libraries etc
- Updates a comment about why we override a method
- Bumps the version
- Tidies up some comments

### CODE REVIEW

File: features/installing_with_cache.feature [hunk 1]
```diff
@@ -23,6 +23,7 @@ Feature: Installing using a compiled-cache
 
     Then I should see "Hello, world!"
 
+  # this works in reality but the test is dodgy on ruby 2. Needs investment in time to fix
```
Comment (caius) [inline features/installing_with_cache.feature]: Raise as an issue?

File: lib/rubygems/commands/precompile_command.rb [hunk 2]
```diff
@@ -11,6 +11,14 @@ def initialize
     add_option('-a','--arch-dirs','Adds the architecture sub-folders to the output directory before writing') do |arch, options|
       options[:arch] = true
     end
+
+    add_option('-n','--noclean','do not clean up tempdirs, leaves temp dirs in place etc for inspection and debugging') do |noclean, options|
+      options[:debug] = true
+    end
+
+    add_option('-b','--build-config=CONFIG','commands to pass to the build i.e the run of extconf.rb') do |config, options|
+      options[:build_config] = config.split(' ')
```
Comment (caius) [inline lib/rubygems/commands/precompile_command.rb]: Would this be better suited to using `Shellwords.split` (for e.g. quoted strings?)

File: lib/rubygems/precompiler.rb [hunk 3]
```diff
@@ -6,16 +6,32 @@
...
     @target_dir = opts.fetch(:output, Dir.pwd)
     @target_dir = File.join(@target_dir, arch_string) if opts.fetch(:arch, false)
+    @debug = opts.fetch(:debug, false)
     @options = opts
+
+    # This writes out a build_info file that the extension builder will process to set the
+    # build configuration. We use the write_build_info_file method which should work on ruby 2+
+    # on ruby  < 1.9.3 we don't support setting of build options. 1.9.3 is EOL.
+    # However most simple gems that do not require build time config will still work.
+    if !opts.fetch(:build_config,[]).empty?
```
Comment (caius) [inline lib/rubygems/precompiler.rb]: I'd be tempted to return early here, there's a fair bit going on inside this nested condition.

```ruby
return if opts.fetch(:build_config, []).empty?

if Gem::Installer.method_defined?…
```

File: lib/rubygems/precompiler.rb [hunk 4]
```diff
@@ -87,6 +107,18 @@ def build_extensions(install_root)
...
+        # the directory we want. Otherwise by default the build process will
+        # look for the gem installed in the usual place won't find it and will then
+        # bail
+        class <<@spec
+          attr_accessor :workroot
+          def full_gem_path
+            return workroot
+          end
+        end
+        @spec.workroot = workroot
```
Comment (caius) [inline lib/rubygems/precompiler.rb]: I _think_ you could replace this with

```ruby
@spec.define_method(:full_gem_path) do
  workroot
end
```

Comment (stevenwilliamson) [inline lib/rubygems/precompiler.rb]: Possible but we also add an attr_accessor, and this version of the code is what is already deployed and tested in production. (It just never got merged to master but we build off a sha). In the interest of knowing it already works and limiting the number of changes id leave it as-is for now. Happy too look at refactoring in a separate release though.

File: features/installing_with_cache.feature (hunk 1 above)
Comment (stevenwilliamson) [inline features/installing_with_cache.feature]: Good idea https://github.com/joshado/rubygems-precompiled/issues/5

File: lib/rubygems/precompiler.rb (hunk 4 above)
Comment (caius) [inline lib/rubygems/precompiler.rb]: :ok_hand:

**Comment by caius** [comment]
Looks good :+1:

**Comment by caius** [comment]
:+1:

QA'd on my build box, does the right thing with the arguments & behaves correctly with uninstalled gems.

**Pull request merged**